kernel/src/fs/utils/id_bitmap.rs

```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::ops::Range;
+
+use aster_block::BLOCK_SIZE;
+use bitvec::{
+    order::Lsb0,
+    slice::BitSlice,
+    view::{AsBits, AsMutBits},
+};
+
+use crate::prelude::*;
+
+/// A disk I/O-friendly bitmap for ID management (e.g., block or inode IDs).
+///
+/// An ID bitmap has the same size as a block, i.e., `BLOCK_SIZE`.
+/// Each bit in the bitmap represents one ID:
+/// bit 0 at the i-th position of the bitmap means the i-th ID is free
+/// and bit 1 means the ID is in use.
+/// As such, the bitmap can contain contain at most `BLOCK_SIZE` * 8 bits/IDs.
+#[derive(Clone)]
+pub struct IdBitmap {
+    buf: Box<[u8]>,
+    first_available_id: u16,
+    len: u16,
+}
+
+impl IdBitmap {
+    /// Creates a new ID bitmap out of a given buffer, whose first `len`-bits represent valid IDs.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if `len` is greater than `IdBitmap::max_len()`.
+    pub fn from_buf(buf: Box<[u8]>, len: u16) -> Self {
+        assert!(len <= Self::max_len());
+        let mut bitmap = Self {
+            buf,
+            first_available_id: 0,
+            len,
+        };
+        bitmap.first_available_id = (0..len)
+            .find(|&i| !bitmap.bit_slice()[i as usize])
+            .unwrap_or(len);
+        bitmap
+    }
+
+    /// Returns the length of the ID bitmap, i.e., the maximum number of IDs.
+    #[expect(unused)]
+    pub const fn len(&self) -> u16 {
+        self.len
+    }
+
+    /// Returns the maximum length of an ID bitmap.
+    pub const fn max_len() -> u16 {
```

Comment:
On a second, calling this method `capacity` would be even more appropriate. Change all occurrences of `max_len`.